File: 189. Rotate Array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int length = nums.size();
        if(k <= 0 || length == 0 || k == length)
            return;
        k %= length;
        reverse(nums, 0, length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, length - 1);
    }
    void reverse(vector<int>& nums, int start, int end){
        if(start >= end)
            return;
        while(start < end){
            swap(nums[start++], nums[end--]);
        }
    }
};

File: 189. Rotate Array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int len = nums.length;
        if(k <= 0 || len == k || len == 0)
            return;
        k %= len;
        reverse(nums, 0, len - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, len - 1);
        
    }
    public static void reverse(int[] nums, int start, int end){
        if(start >= end)
            return;
        while(start < end){
            int temp = nums[start];
            nums[start++] = nums[end];
            nums[end--] = temp;
        }
    }
}